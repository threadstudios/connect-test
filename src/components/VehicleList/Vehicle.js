import Inferno from 'inferno';
import fetch from 'isomorphic-fetch';
import Loading from '../Loading/Loading';
import ErrorComponent from '../Error/Error';
import Image from '../PreloadedImage';
import {compose, branch, withLifecycle, withState, defaultProps, renderComponent} from 'incompose';

const Vehicle = (props) => {
    return (
        <div className={props.className}>
            <Image className="vehicle_image" url={props.media[0].url} />
            <div className="vehicle_description">
                <h1><span>{props.id.toUpperCase()}</span></h1>
                <h2>From {props.data.price}</h2>
                <p>{props.data.description}</p>
            </div>
        </div>
    );
}

const fetchData = (id) => {
    const apiUrl = process.env.API_URL ? process.env.API_URL : __CONFIG__.API_URL;
    return new Promise((resolve, reject) => {
        fetch(`${apiUrl}api/vehicle/${id}`)
        .then((result) => {
            return result.json();
        })
        .then((json) => {
            return resolve(json);
        })
        .catch((err) => {
            return reject(err);
        });
    })
}

const withLoadingComponent = branch(
    (props) => props.loading,
    renderComponent(Loading)
);

const withErrorComponent = branch(
    (props) => props.error,
    renderComponent(ErrorComponent)
)

const withFetcher = branch(
    (props) => !props.data,
    withLifecycle({
        componentDidMount() {
            this.setState({loading: true});
            fetchData(this.props.id)
            .then((result) => {
                this.setState({
                    loading: false,
                    data : result
                })
            })
            .catch((err) => {
                this.setState({
                    loading: false,
                    error: true,
                    msg: err.message
                })
            });
        }
    })
)   

export default compose(
    withState('data', 'setData', false),
    withState('loading', 'setLoading', true),
    withState('error', 'setError', false),
    withState('msg', 'setErrorMsg', false),
    defaultProps({ className : 'vehicle' }),
    withFetcher,
    withLoadingComponent,
    withErrorComponent
)(Vehicle)