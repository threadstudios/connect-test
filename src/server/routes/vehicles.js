import { Router } from 'express';
import vehicles from 'technical-test/server_api/vehicles';

const api = new Router();

api.use((req, res, next) => {
    req.store = vehicles;
    return next();
});

api.get('/api/vehicle', (req, res) => {
    return res.json(vehicles);
});

api.get('/api/vehicle/:id', (req, res) => {
    // clean the id
    let id = req.params.id.replace(/[^\w\s\-]*/, "");
    try {
        let vehicle = require(`technical-test/server_api/vehicle_${id}`);
        return res.json(vehicle);
    } catch(e) {
        return res.status(400).json({
            error: 'Vehicle could not be found'
        })
    }
});

export default api;