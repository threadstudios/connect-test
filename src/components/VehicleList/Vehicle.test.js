import Inferno from 'inferno';
import { renderToSnapshot, renderIntoDocument, findRenderedDOMElementWithClass, findRenderedVNodeWithType } from 'inferno-test-utils';
import Vehicle from './Vehicle';
import vehicles from 'technical-test/server_api/vehicles';
import individualVehicle from 'technical-test/server_api/vehicle_xe';

describe("A Vehicle Component", () => {

    const vehicleData = vehicles.vehicles[0];

    window.__CONFIG__ = {API_URL : '/'}

    it("should render without an error", () => {
        expect(renderToSnapshot(<Vehicle />)).toMatchSnapshot();
    })

    it("should display a loader before the vehicle is displayed", () => {
        const wrapper = renderIntoDocument(<Vehicle className="foo" {...vehicleData} />);
        const result = findRenderedDOMElementWithClass(wrapper, "loading");
        expect(result).toBeDefined();
    })

    it("should display the vehicle information when it is fetched", (done) => {

        const wrapper = renderIntoDocument(<Vehicle className="foo" {...vehicleData} />);
        const vehicle = findRenderedVNodeWithType(wrapper, Vehicle);

        // Set the state to be loaded
        vehicle.children.updateStateValue({
            data : individualVehicle,
            loading: false
        }, () => {
            expect(renderToSnapshot(vehicle)).toMatchSnapshot();
            done();
        });

    })

})