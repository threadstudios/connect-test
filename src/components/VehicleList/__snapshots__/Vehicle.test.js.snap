// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A Vehicle Component should display the vehicle information when it is fetched 1`] = `
<div
  className="foo"
  id="xe"
  media={
    Array [
      Object {
        "name": "vehicle",
        "url": "/images/xe_k17.jpg",
      },
    ]
  }
  modelYear="k17"
  url="/api/vehicle/xe"
/>
`;

exports[`A Vehicle Component should render without an error 1`] = `<div />`;
